app/src/main/java/ch/epfl/qedit/backend/database/CacheDBService.java

```diff
@@ -0,0 +1,226 @@
+import java.io.ObjectInputStream;
+import java.io.ObjectOutputStream;
+import java.io.Serializable;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.CompletableFuture;
+import java.util.function.Function;
+import java.util.function.Supplier;
+
+/**
+ * This class decorates a database service so that some of the requests are cached on the internal
+ * storage. This is used to reduce the delay of accessing the database and allowing the user to use
+ * the app without internet connection.
+ */
+public class CacheDBService implements DatabaseService {
+
+    // This is the underlying database service that is being decorated
+    private final DatabaseService dbService;
+
+    // These are the directories in which we cache the data
+    private final File userCacheDir;
+    private final File quizCacheDir;
+    private final File poolCacheDir;
+    private final File langCacheDir;
+
+    /** Creates a cache that decorates the given database service */
+    public CacheDBService(DatabaseService dbService, Context context) {
+        this.dbService = dbService;
+        this.userCacheDir = new File(context.getCacheDir(), "users");
+        this.quizCacheDir = new File(context.getCacheDir(), "quizzes");
+        this.poolCacheDir = new File(context.getCacheDir(), "pools");
+        this.langCacheDir = new File(context.getCacheDir(), "languages");
+
+        // Create the subdirectories if they do not already exist
+        createDir(userCacheDir);
+        createDir(quizCacheDir);
+        createDir(poolCacheDir);
+        createDir(langCacheDir);
+    }
+
+    private static void createDir(File dir) {
+        if (!dir.exists()) dir.mkdirs();
+    }
+
+    @Override
+    public CompletableFuture<List<String>> getQuizLanguages(String quizId) {
+        // Lists are not serializable, but array lists are, so we just need to wrap the lists
+        // into array lists before storing them in the cache. Since array lists are lists,
+        // nothing needs to be done to deserialize.
+        return this.<List<String>, ArrayList<String>>retrieve(
+                new File(langCacheDir, quizId),
+                () -> dbService.getQuizLanguages(quizId),
+                list -> new ArrayList<>(list),
+                list -> list // cannot use Function.identity() because of the type system
+                );
+    }
+
+    @Override
+    public CompletableFuture<Quiz> getQuizStructure(String quizId) {
+        return retrieve(new File(quizCacheDir, quizId), () -> dbService.getQuizStructure(quizId));
+    }
+
+    @Override
+    public CompletableFuture<StringPool> getQuizStringPool(String quizId, String language) {
+        return retrieve(
+                new File(poolCacheDir, quizId + "-" + language),
+                () -> dbService.getQuizStringPool(quizId, language));
+    }
+
+    @Override
+    public CompletableFuture<String> uploadQuiz(Quiz quiz, StringPool stringPool) {
```

Comment:
So then one would store the quiz for instance manually using `storeInCache(...)` before calling this method that uploads the quiz to the real database. This allows more flexibility when deciding what is going to be cached and what not. At first I thought that we might always put the quiz in the cache here as well before uploading to the real database. But as I said this would be less flexible. Did I understand your intentions correctly?

Reply:
I didn't plan on manually storing the quizzes in the cache, because it would spread the logic to other places in the code. The problem that I though of is that if you upload a quiz and then share it using a QR code, you need the quiz to be loaded in the database. That's why I always upload it for now.
So indeed, what you said would be the only solution to our problem...